### PR TITLE
CLN: trips as geodataframe with start and end coordinates

### LINKED ISSUE #57: start and endpoint for trip model
I think trips in trackintel should have the start and end point as geometry. Even though from the perspective of creating trips from staypoints this is redundant but in real-life use cases it might appear that we only analyze trips without having access to staypoints.

- Add columns geom_start and geom_end as required (and with a fixed name I guess...)
- Add column primary_mode as an optional column

### PR DESCRIPTION
**Pull request by NinaWie**
Closes #57 
The aim was to add start and end point of trips to the trips data frame. This however required major changes to the tests, since so far the trips data frames are assumed to be dataframes in trackintel, and not geodataframes. An alternative to changing the test data would be given by keeping the compatibility with data frames, e.g. adding the following line in read and write functions:
```
if "geom" in df.columns:
    df["geom"] = df["geom"].apply(wkt.dumps)
```

Also, problems occur if start or end activity of a trip are not given, because MultiPoint cannot contain null points or empty points. I thus suggest to set the start coordinates of the trip to the start of its first tripleg (and same for the destination coordinates, to the last coordinates on the last tripleg) as an approximation. Happy for other suggestions though.

### CODE REVIEW

File: trackintel/io/file.py
Comment (NinaWie) [inline trackintel/io/file.py:476]: Here one would have to add an if clause to keep the compatibility with dataframes (compared to geodataframes)

File: tests/data/trips_mod_columns.csv
Comment (NinaWie) [inline tests/data/trips_mod_columns.csv:1]: I left the geom column here in order to test whether the dataframes are equal (except for this column) - Tested in `test_file.py`. Apart from this one and the `geolife_long/trips.csv`, I changed the test data back to dataframes, since trips tables now support either dataframes or geodataframes.

File: trackintel/model/trips.py
Comment (NinaWie) [inline trackintel/model/trips.py:28]: I removed this because I think it is wrong, or misleading. As far as I could see, the triplegs before or after the first or last activity are also considered as trips, and are not ommitted. Please correct me if I'm wrong there

**Comment by henrymartin1** [comment]
Results from our last meeting: 
- Geometry should be optional
- Multipoint with the first point of the first tripleg (or the last point of the last tripleg) can be used if the start/end point is missing because multipoint does not allow for missing geometries

File: tests/data/geolife_long/trips.csv
Comment (henrymartin1) [inline tests/data/geolife_long/trips.csv:2]: If the geometry is optional there should be some tests without geometry to see if that works correctly

File: tests/data/trips_mod_columns.csv
Comment (henrymartin1) [reply tests/data/trips_mod_columns.csv:1]: Ah ok. So to have some test data with and some without geometry?

File: tests/io/test_file.py
Comment (henrymartin1) [inline tests/io/test_file.py:289]: We should talk about accessors ;-)

File: tests/preprocessing/test_triplegs.py
Comment (henrymartin1) [inline tests/preprocessing/test_triplegs.py:79]: Change comment to destination

File: tests/preprocessing/test_triplegs.py
Comment (henrymartin1) [inline tests/preprocessing/test_triplegs.py:70]: I think this should be its own test.

File: trackintel/io/file.py
Comment (henrymartin1) [inline trackintel/io/file.py:427]: Here we should describe that the type is "multipoint"

File: trackintel/io/file.py
Comment (henrymartin1) [inline trackintel/io/file.py:438]: I think the return type should be unique

File: trackintel/io/file.py
Comment (henrymartin1) [reply trackintel/io/file.py:427]: I would be good if you could update the example under "examples" in the docstring below

File: trackintel/io/file.py
Comment (henrymartin1) [inline trackintel/io/file.py:479]: if we want a unique return type, then it should always be transformed into a geodataframe

File: trackintel/model/trips.py
Comment (henrymartin1) [inline trackintel/model/trips.py:22]: I think the geometry column could be added here. Additionally I would add the sentence "Trips have an optional geometry of type multipoint which describes the start and the end point of the trip" or something similar.

This docstring here is very important because it serves as the main documentation for trips.

File: trackintel/model/trips.py
Comment (henrymartin1) [reply trackintel/model/trips.py:28]: No, I think you are right. It was different in the past, thanks for the update

File: trackintel/model/trips.py
Comment (henrymartin1) [inline trackintel/model/trips.py:34]: typo in "unkown"

File: trackintel/model/trips.py
Comment (henrymartin1) [reply trackintel/model/trips.py:22]: The sentence I proposed could also be added to "notes"

File: trackintel/preprocessing/triplegs.py
Comment (henrymartin1) [inline trackintel/preprocessing/triplegs.py:236]: Oh shit... this will be horrible to merge with our updated trip generation [PR220](https://github.com/mie-lab/trackintel/pull/220)
The best thing to do might be to wait until we merged it... Unfortunatly it is somewhat stuck with a bug we are working on...

File: trackintel/model/trips.py
Comment (henrymartin1) [inline trackintel/model/trips.py:69]: nice!

File: trackintel/model/trips.py
Comment (henrymartin1) [inline trackintel/model/trips.py:75]: Can you add a test for that special case? Can be a small one with a dummy example

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
Great work! It is honestly a complicated issue because it touches on the core of the data model. 
Special credit for reading through the trip generation! I hope you still like trackintel ;-) 

Here are a few additional comments: 
- Is it possible to have it always as a Geodataframe even though it sometimes is without geometries? I think that would be cleaner than if it always changes types
- We should have a test that checks if the behavior for taking the first point of a tripleg or the last point of a staypoint is correct. 

The generate_trips function is under review at the moment and will receive a major update (#220) we should discuss on Wednesday what we merge first.

**Comment by henrymartin1** [comment]
Reminder: In the end we will have to update the documentation

**Comment by NinaWie** [comment]
I changed the structure according to your comment @henrymartin1 such that the trips is always a GeoDataFrame, no matter if it contains a geometry column or not. However, I get an error and I'm confused about this test: https://github.com/mie-lab/trackintel/blob/e0c0cdd0d8472ba7b113b3819d062ea8abcd8168/tests/io/test_from_geopandas.py#L64
In all other `from_geopandas` tests, the reading from a .geojson file is compared to reading form a csv file. In contrast, in the trips tests the csv file is basically compared to a csv. Since one of them is a GeoDataFrame now, the test fails. 

I could now either 1) change the test such that the first one is also a gdf, which however makes the test quite useless (because we load the same csv file in both cases and then convert it to a gdf), or 2) generate a new geojson file from the `trips.csv` (which does not have a geometry column) or `geolife_long/trips.csv` (which has a geometry column). 
Let me know what you think.

File: tests/data/geolife_long/trips.csv
Comment (NinaWie) [reply tests/data/geolife_long/trips.csv:2]: Yes, these tests exist in `io/test_file.py`

File: trackintel/io/file.py
Comment (NinaWie) [reply trackintel/io/file.py:427]: Changed this now

File: trackintel/model/trips.py
Comment (NinaWie) [reply trackintel/model/trips.py:22]: Also changed with the last commits

File: trackintel/model/trips.py
Comment (NinaWie) [reply trackintel/model/trips.py:74]: I added this test in `tests/model/test_trips.py`, see below

File: trackintel/preprocessing/triplegs.py
Comment (NinaWie) [reply trackintel/preprocessing/triplegs.py:236]: Yes let's discuss this on Wednesday

File: tests/io/test_from_geopandas.py
Comment (NinaWie) [inline tests/io/test_from_geopandas.py:72]: I had to add this line to make the test work (since `read_trips_csv` now always returns a GeoDataFrame), but as I commented, the test is rather useless now. Maybe we should add a geojson for `trips.csv`.

File: tests/data/trips_mod_columns.csv
Comment (NinaWie) [reply tests/data/trips_mod_columns.csv:1]: Yes, as said above, in `test_file.py` some tests read csv files with geometry and others without, and the two are compared as well for consistency

File: tests/preprocessing/test_triplegs.py
Comment (NinaWie) [inline tests/preprocessing/test_triplegs.py:209]: Here, we could also change the `_old_generate_trips` function to include the geometry, but it seems a bit overkill since the function is deprecated anyway. However, it would then of course be tested, whether iterating for each user leads also to the same geometry column as the function implemented now (at the moment I am sure that it yields the same result, since the test data was generated with my previous iterative version). Let me know if you think we should modify the `_old_generate_trips`.

File: trackintel/preprocessing/triplegs.py
Comment (NinaWie) [inline trackintel/preprocessing/triplegs.py:115]: concat does not work for geometries

File: trackintel/preprocessing/triplegs.py
Comment (NinaWie) [inline trackintel/preprocessing/triplegs.py:220]: I did use the triplegs id now to join the two tables, in order to fill in the points where the staypoint is NaN. Maybe there is a smarter way to do this.

**Review by NinaWie — COMMENTED** [review]
I wondered whether we should make the `geom` column optional in `generate_trip`, i.e. as a kw argument. This would avoid the computational effort of getting the geometry from stps and tpls if the user does not need it. Let me know if you think we should change that.

File: tests/io/test_from_geopandas.py
Comment (henrymartin1) [reply tests/io/test_from_geopandas.py:72]: I am not so sure if I understand what you mean.
Does that mean, that `read_trips_gpd` does not return a geodataframe?

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
Great work! Especially because this is quite a complex PR that affects so many different areas of the library. It seems like you know trackintel quite well now ;-)
I have added some comments in the file and I have two general comments:
- Be aware that any input geodataframe can have arbitrary geometry column names (e.g., spts/tpls in the generate_trips function.
- Please add a bullet point in the notes section of the docstring of the generate_trips function that says something like "trips optionally have their start and end point as geometry of type multipoint". You should also describe in the docstring that we take the first point of the first tripleg in case of missing origin/destination and that these trips can still be found because they miss the orgin/destination id. (Maybe its rather a small subsection than a bulletpoint ;-) )
- I think it would be a good idea to make the calculation of the geometry optional to have a faster version of the trip_generation if needed (if it is not too much work). I think the default should be that the geometry is generated

File: trackintel/model/trips.py
Comment (NinaWie) [reply trackintel/model/trips.py:71]: At the moment, checking `obj.geometyr is None` only leads to an error `AttributeError: No geometry data set yet`. Furthermore, it seems that geopandas will at some point be updated to return a dataframe if no geometry column exists (https://github.com/geopandas/geopandas/issues/1911). This is problematic for us, since the code now is designed in a way that the geometry column is optional, but trips is still always a GeoDataFrame. Since this is already tricky, and will probably become impossible with new geopandas versions, I'd maybe suggest to go back to the option we previously considered, where trips can be either a GeoDataframe or a DataFrame? Let me know what you think.

Comment (henrymartin1) [reply trackintel/model/trips.py:71]: Ah, I see. Yes in this case it is better to go for your initial idea. In the case it is a geodataframe we still need to address the geometry column dynamically.

File: tests/preprocessing/test_triplegs.py
Comment (NinaWie) [inline tests/preprocessing/test_triplegs.py:90]: I added this test to compare the result with and without computing the geometry

File: tests/preprocessing/test_triplegs.py
Comment (NinaWie) [inline tests/preprocessing/test_triplegs.py:356]: The optional `add_geometry` actually simplified some tests; here for example we don't need to drop the geometry afterwards.

File: trackintel/io/file.py
Comment (NinaWie) [inline trackintel/io/file.py:509]: I am now using `isinstance` in all cases to check whether the geometry exists or not. We have to be careful with GeoDataFrames that do not contain a geometry in further PRs

File: trackintel/preprocessing/triplegs.py
Comment (NinaWie) [inline trackintel/preprocessing/triplegs.py:89]: Added this to the notes in response to your comment

File: trackintel/preprocessing/triplegs.py
Comment (NinaWie) [inline trackintel/preprocessing/triplegs.py:248]: I liked your solution better, and I think it was only unreadable because of the `trip_idx` variable etc that I had earlier. Now, I got rid of the nested function and simply repeat the same step for the distination.

**Review by NinaWie — COMMENTED** [review]
Should be much cleaner now, thanks for your comments! Let me know what else is left to do

File: tests/io/test_from_geopandas.py
Comment (NinaWie) [reply tests/io/test_from_geopandas.py]: Never mind, this changed again with the last commit :)

File: trackintel/model/trips.py
Comment (henrymartin1) [inline trackintel/model/trips.py:38]: [Doc]
"geometry" needs to start exactly below "if" 
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#lists-and-quote-like-blocks

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
Hey Nina, 
great work! I think everything concerning content is done. 🎉 I had several small minor comments concerning the clearity of the comments and the documentation. Usually I would simply add them myself and then merge but I had one question in ` tests/model/test_trips.py ` where I was not sure if it was a typo or not .
For the other comments I have added what I would have changed as comments, you can either include this directly or I can do it before I merge it.


[Documentation]
At the end of an important pull request it is often a good idea to compile the documentation and check if everything is set correctly. [Here](https://github.com/mie-lab/trackintel/blob/master/CONTRIBUTING.md#documentation) is a bad description on how to do it. After you installed sphinx I think you also need to install "sphinx_rtd_theme" but I am not sure...
It is not really necessary that you do that now everything works fine if you fix the number of spaces in the docstring of `TripsAccessor`.

**Review by hong2223 — APPROVED** [review]

**Comment by NinaWie** [comment]
Thanks for the review; I did the changes to the documentation as you suggested :) I also compiled it locally and - after removing some more spaces - it looked fine. Let me know if there is anything else to do

**Review by henrymartin1 — APPROVED** [review]
Nice work, I am very happy to merge this now. Congratulations and thank you for your first major contribution to the framework!

**Pull request merged**
